ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -80,66 +86,97 @@ public void run() {
             timer.schedule(refreshIngestionResourceValuesTask, 0, REFRESH_INGESTION_RESOURCES_PERIOD);
 
         } catch (Exception e) {
-            log.error(String.format("Error in initializing ResourceManager: %s.", e.getMessage()), e);
+            log.error("Error in initializing ResourceManager: {}.", e.getMessage(), e);
             throw e;
         }
     }
 
-    public void clean() {
+    void clean() {
         ingestionResources.clear();
     }
 
-    public String getKustoIdentityToken() throws Exception {
-        if (identityToken == null) {
-            refreshIngestionAuthToken();
-            if (identityToken == null) {
-                throw new Exception("Unable to get Identity token");
+    String getIngestionResource(ResourceTypes resourceType) throws Exception {
+        if (!ingestionResources.containsKey(resourceType)) {
+            // When the value is not available, we need to get the tokens from Kusto (refresh):
+            refreshIngestionResources();
+            try{
+                // If the write lock is locked, then the read will wait here.
+                // In other words if the refresh is running yet, then wait until it ends:
+                ingestionResourcesLock.readLock().lock();
+                if (!ingestionResources.containsKey(resourceType)) {
+                    throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+                }
+            } finally {
+                ingestionResourcesLock.readLock().unlock();
             }
         }
-        return identityToken;
+        return ingestionResources.get(resourceType).nextValue();
     }
 
-    public String getIngestionResource(ResourceTypes resourceType) throws Exception {
-        if (!ingestionResources.containsKey(resourceType)) {
-            refreshIngestionResources();
-            if (!ingestionResources.containsKey(resourceType)) {
-                throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+    String getKustoIdentityToken() throws Exception {
+        if (identityToken == null) {
+            refreshIngestionAuthToken();
```

Comment:
In the current implementation, for every ask of the token, we'll try to acquire a write token. 
Isn't is better to first check if we need to refresh and only then try to acquire the lock?
There is the Double-checked locking (if lock if) pattern for that.

Reply:
Why would the token be null unless it's the first refresh?